SimPEG/DCIP/BaseIP.py

```diff
@@ -0,0 +1,182 @@
+        return self._dMdsig
+
+    @property
+    def A(self):
+        """
+            Makes the matrix A(m) for the DC resistivity problem.
+
+            :param numpy.array m: model
+            :rtype: scipy.csc_matrix
+            :return: A(m)
+
+            .. math::
+                c(m,u) = A(m)u - q = G\\text{sdiag}(M(mT(m)))Du - q = 0
+
+            Where M() is the mass matrix and mT is the model transform.
+        """
+        if getattr(self, '_A', None) is None:
+            D = self.mesh.faceDiv
+            G = self.mesh.cellGrad
+            self._A = D*self.Msig*G
+            # Remove the null space from the matrix.
+            self._A[-1,-1] /= self.mesh.vol[-1]
+            self._A = self._A.tocsc()
+        return self._A
+
+    def getRHS(self):
+        # if self.mesh not in self._rhsDict:
+        RHS = np.array([src.eval(self) for src in self.survey.srcList]).T
+        # self._rhsDict[mesh] = RHS
+        # return self._rhsDict[mesh]
+        return RHS
+
+    def fields(self, m):
+        if self.u is None:
+	        A    = self.A
+	        if self.Ainv == None:
+	        	self.Ainv = self.Solver(A, **self.solverOpts)
+	        Q    = self.getRHS()
+	        self.u  = self.Ainv * Q
+        return self.u
+
+    def forward(self, m, u=None):
+        # Set current model; clear dependent property $\mathbf{A(m)}$
+        self.curModel = m
+        # sigma = self.curModel.transform # $\sigma = \mathcal{M}(\m)$
+        sigma = self.sigma
+        if self.u is None:
+            # Run forward simulation if $u$ not provided
+            u = self.fields(sigma)
+
+        shp = (self.mesh.nC, self.survey.nSrc)
+        u = self.u.reshape(shp, order='F')
+
+        D = self.mesh.faceDiv
+        G = self.mesh.cellGrad
+        # Derivative of model transform, $\deriv{\sigma}{\m}$
+        # dsigdm_x_v = self.curModel.transformDeriv * v
+
+        dsigdm_x_v = Utils.sdiag(sigma) * self.curModel.transformDeriv *  m
+
+        # Take derivative of $C(m,u)$ w.r.t. $m$
+        dCdm_x_v = np.empty_like(u)
+        # loop over fields for each source
+        for i in range(self.survey.nSrc):
+            # Derivative of inner product, $\left(\mathbf{M}_{1/\sigma}^f\right)^{-1}$
+            dAdsig         = D * self.dMdsig( G * u[:,i] )
+            dCdm_x_v[:, i] = dAdsig *  dsigdm_x_v
+
+        # Take derivative of $C(m,u)$ w.r.t. $u$
+
+        if self.Ainv == None:
+        	self.Ainv = self.Solver(A, **self.solverOpts)
+
+        # dCdu = self.A
+        # Solve for $\deriv{u}{m}$
+        # dCdu_inv = self.Solver(dCdu, **self.solverOpts)
+        P        = self.survey.getP(self.mesh)
+        J_x_v    = - P * mkvc( self.Ainv * dCdm_x_v )
+        return -J_x_v
+
+    def Jvec(self, m, v, u=None):
+        return self.forward(v)
+
+    def Jtvec(self, m, v, u=None):
+
+        self.curModel = m
+        # sigma = self.curModel.transform # $\sigma = \mathcal{M}(\m)$
+        sigma = self.sigma
+        if self.u is None:
```

Comment:
this should be `if u is None`? (`u` is an arg passed to Jtvec, it should not be assigned to self?)